File: MinStack/min_stack.go
package min_stack

import "math"

type MinStack struct {
	stack      []int
	minElement int
}

func Constructor() MinStack {
	minStack := MinStack{stack: []int{}, minElement: math.MaxInt64}
	return minStack
}

func (this *MinStack) Push(val int) {
	this.stack = append(this.stack, val)

	if val < this.minElement {
		this.minElement = val
	}
}

func (this *MinStack) Pop() {
	popElement := this.Top()
	this.stack = this.stack[:len(this.stack)-1]

	if this.minElement == popElement {
		this.minElement = math.MaxInt64
		for _, v := range this.stack {
			if v < this.minElement {
				this.minElement = v
			}
		}
	}
}

func (this *MinStack) Top() int {
	return this.stack[len(this.stack)-1]
}

func (this *MinStack) GetMin() int {
	return this.minElement
}

File: MinStack/min_stack.js
var MinStack = function () {
	this.stack = [];
	this.minStack = [];
};

/**
 * @param {number} val - Add value to stack
 * @returns {void}
 */
MinStack.prototype.push = function (val) {
	this.stack.push(val);

	if (
		this.minStack.length === 0 ||
		this.minStack[this.minStack.length - 1] >= val
	) {
		this.minStack.push(val);
	}
};

/**
 * @returns {void}
 */
MinStack.prototype.pop = function () {
	if (this.stack.length !== 0) {
		if (
			this.stack[this.stack.length - 1] ===
			this.minStack[this.minStack.length - 1]
		) {
			this.minStack.pop();
		}

		this.stack.pop();
	}
};

/**
 * @returns {number}
 */
MinStack.prototype.top = function () {
	return this.stack[this.stack.length - 1];
};

/**
 * @returns {number}
 */
MinStack.prototype.getMin = function () {
	return this.minStack[this.minStack.length - 1];
};
